### PR TITLE
Restart pub/sub server on service reset

### PR DESCRIPTION
**Pull request by simonvadee**
Hi fellas,

I encountered an issue when stopping/restarting a `rpc.client.Client` because of:
```golang
// Start implements Service by calling OnStart (if defined). An error will be
// returned if the service is already running or stopped. Not to start the
// stopped service, you need to call Reset.
func (bs *BaseService) Start() error { ... }
```

Therefore I used `func (bs *BaseService) Reset()` but since the `EventBus` does not implement `OnReset()` I got a `PanicSanity("The service cannot be reset")`. This is why I added this method to `EventBus`.

The reason why I want to reset the service is that once it is stopped, the pub/sub server is shutdown and cannot be restarted, so even if subscribers are still able to subscribe to new events, they are not notified.

I don't really know what kind of tests you'd like me to write or doc you want me to update so let me know what you'd like !

Cheers!

### CODE REVIEW

File: types/event_bus.go [hunk 1]
```diff
@@ -52,6 +52,10 @@ func (b *EventBus) OnStop() {
 	b.pubsub.OnStop()
 }
 
+func (b *EventBus) OnReset() error {
+	return b.pubsub.OnStart()
```
Comment (melekes) [inline types/event_bus.go]: why not `OnReset`?

Comment (simonvadee) [reply types/event_bus.go]: `pubsub.Sever` does not implement `OnReset()` and `OnStart()` starts the server as intended, do you want me to add the missing method in `tendermint/tmlibs` ? And then should I point glide.yaml to the master branch of tmlibs ?

Comment (melekes) [reply types/event_bus.go]: > do you want me to add the missing method in tendermint/tmlibs ?

yes, please

```go
func (s *Server) OnReset() error {	return nil }
```

> And then should I point glide.yaml to the master branch of tmlibs ?

no, point it to commit in tmlibs

Comment (melekes) [reply types/event_bus.go]: Want me to do it?

Comment (simonvadee) [reply types/event_bus.go]: It's not an urgent matter, i can take care of it when I'll have a few minutes, but thank's for asking :)

**Comment by melekes** [comment]
merged associated PR. @simonvadee could you update this one when u have time?

**Comment by melekes** [comment]
Hmmm.. tests are failing. Could you also rebase your code against latest `develop`?

**Comment by xla** [comment]
@melekes The tests timed out, kicked off a re-run.

**Comment by xla** [comment]
@simonvadee We did some changes to our CI, could you rebase latest `tendermint:develop`?

**Comment by melekes** [comment]
> https://github.com/tendermint/tendermint/pull/1239#issuecomment-370763774

did not help.

**Comment by simonvadee** [comment]
what's the status on this, do you want me to do something about the tests ?

**Comment by xla** [comment]
@simonvadee Looks like the change you made introduced a condition where our tests are blocked, this needs further investigation. Maybe you can reproduce it locally as a start?

**Comment by ebuchman** [comment]
Thanks for this.

Unfortunately there have been a number of issues around Restart and we'll probably be getting rid of it altogether. Instead of restarting, one should just recreate the service and call Start. The `Service` is undergoing refactor and will move into tendermint. See https://github.com/tendermint/tmlibs/pull/114

**Pull request closed without merging**